fdbserver/RestoreMaster.actor.h

```diff
@@ -90,98 +101,188 @@ struct RestoreMasterData : RestoreRoleData, public ReferenceCounted<RestoreMaste
+	std::tuple<double, int, std::vector<RestoreFileFR>> getVersionSize(Version prevVersion, Version nextVersion,
+	                                                                   const std::vector<RestoreFileFR>& rangeFiles,
+	                                                                   int rangeIdx,
+	                                                                   const std::vector<RestoreFileFR>& logFiles) {
+		double size = 0;
+		TraceEvent("FastRestoreGetVersionSize")
+		    .detail("PreviousVersion", prevVersion)
+		    .detail("NextVersion", nextVersion)
+		    .detail("RangeFiles", rangeFiles.size())
+		    .detail("RangeIndex", rangeIdx)
+		    .detail("LogFiles", logFiles.size());
+		ASSERT(prevVersion <= nextVersion);
+		while (rangeIdx < rangeFiles.size()) {
+			TraceEvent(SevDebug, "FastRestoreGetVersionSize").detail("RangeFile", rangeFiles[rangeIdx].toString());
+			if (rangeFiles[rangeIdx].version < nextVersion) {
+				ASSERT(rangeFiles[rangeIdx].version >= prevVersion);
+				size += rangeFiles[rangeIdx].fileSize;
 			} else {
-				vbIter->second.logFiles.push_back(allFiles[i]);
+				break;
 			}
+			++rangeIdx;
 		}
-
-		// Sort files in each of versionBatches and set fileIndex, which is used in deduplicating mutations sent from
-		// loader to applier.
-		// Assumption: fileIndex starts at 1. Each loader's initized fileIndex (NotifiedVersion type) starts at 0
-		int fileIndex = 0; // fileIndex must be unique; ideally it continuously increase across verstionBatches for
-		                   // easier progress tracking
-		int versionBatchId = 1;
-		for (auto versionBatch = versionBatches->begin(); versionBatch != versionBatches->end(); versionBatch++) {
-			std::sort(versionBatch->second.rangeFiles.begin(), versionBatch->second.rangeFiles.end());
-			std::sort(versionBatch->second.logFiles.begin(), versionBatch->second.logFiles.end());
-			for (auto& logFile : versionBatch->second.logFiles) {
-				logFile.fileIndex = ++fileIndex;
-				TraceEvent("FastRestore")
-				    .detail("VersionBatchId", versionBatchId)
-				    .detail("LogFile", logFile.toString());
-			}
-			for (auto& rangeFile : versionBatch->second.rangeFiles) {
-				rangeFile.fileIndex = ++fileIndex;
-				TraceEvent("FastRestore")
-				    .detail("VersionBatchId", versionBatchId)
-				    .detail("RangeFile", rangeFile.toString());
+		int logIdx = 0;
+		std::vector<RestoreFileFR> retLogs;
+		// Scan all logFiles every time to avoid assumption on log files' version ranges.
+		// For example, we do not assume each version range only exists in one log file
+		while (logIdx < logFiles.size()) {
+			Version begin = std::max(prevVersion, logFiles[logIdx].beginVersion);
+			Version end = std::min(nextVersion, logFiles[logIdx].endVersion);
+			if (begin < end) { // logIdx file overlap in [prevVersion, nextVersion)
+				double ratio = (end - begin) / (logFiles[logIdx].endVersion - logFiles[logIdx].beginVersion);
```

Comment:
`ratio` will always be a whole number here. You need to cast at least one operand of `/` to `double` in order to call the correct operator.